crates/lox_io/src/spice.rs

```diff
@@ -0,0 +1,369 @@
+use nom::branch::alt;
+use nom::bytes::complete::{tag, take_until, take_while1};
+use nom::character::complete::{alpha1, line_ending, multispace0, one_of};
+use nom::combinator::{map, map_res, recognize, rest};
+use nom::multi::{fold_many0, fold_many1, many0, many1};
+use nom::number::complete::{double, float};
+use nom::sequence::{delimited, preceded, separated_pair, terminated, tuple};
+use nom::IResult;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq)]
+enum Value {
+    Double(f64),
+    String(String),
+    DoubleArray(Vec<f64>),
+    StringArray(Vec<String>),
+}
+
+pub struct Kernel {
+    type_id: String,
+    items: HashMap<String, Value>,
+}
+
+type Entries = Vec<(String, Value)>;
+
+impl Kernel {
+    pub fn parse(input: &str) -> Result<Self, &'static str> {
+        if let Ok(("", (type_id, entries, _))) = kernel(input) {
+            Ok(Self {
+                type_id: type_id.to_string(),
+                items: entries.into_iter().collect(),
+            })
+        } else {
+            Err("kernel parsing failed")
+        }
+    }
+
+    pub fn type_id(&self) -> &str {
+        &self.type_id
+    }
+
+    pub fn get_double(&self, key: &str) -> Option<f64> {
+        let value = self.items.get(key)?;
+        if let Value::Double(v) = value {
+            Some(*v)
+        } else {
+            None
+        }
+    }
+
+    pub fn get_double_array(&self, key: &str) -> Option<&Vec<f64>> {
+        let value = self.items.get(key)?;
+        if let Value::DoubleArray(v) = value {
+            Some(v)
+        } else {
+            None
+        }
+    }
+}
+
+fn kernel(s: &str) -> IResult<&str, (&str, Entries, &str)> {
+    let mut parser = tuple((
+        header,
+        fold_many0(
+            preceded(take_until("\\begindata\n"), data_block),
+            Vec::new,
+            |mut out: Entries, item: Entries| {
+                out.extend(item);
+                out
+            },
+        ),
+        rest,
+    ));
+    parser(s)
+}
+
+fn header(s: &str) -> IResult<&str, &str> {
+    let mut parser = preceded(tag("KPL/"), alpha1);
+    parser(s)
+}
```

Comment:
Isn'a syntax like this available:

```
let header = preceded(...);
```

It would get rid of this extra function wrapper which you don't need

Reply:
Done